feat: introduce offshore wind hubs topology

tgilon opened the pull request:
Closes https://github.com/open-energy-transition/open-tyndp/issues/48. This PR builds on top of https://github.com/open-energy-transition/open-tyndp/pull/53.

## Changes proposed in this Pull Request

This PR introduces the topology of offshore wind hubs. It reads in the [Offshore Hub Modelling Inputs](https://2024-data.entsos-tyndp-scenarios.eu/files/scenarios-inputs/Offshore-hubs.zip) from the [TYNDP 2024](2024.entsos-tyndp-scenarios.eu). The grid topology (buses coordinates and transmission capacities) is exogenously set using these data.

The following schematic illustrates an overview of the 2024 TYNDP Offshore Hub topology as implemented in this PR:
![open-tyndp concepts - TYNDP Offshore topology](https://github.com/user-attachments/assets/cbb0e418-6c22-438b-81d2-ed437c80f90c)

## Tasks
- [x] Prepare reference grid data set
- [x] Add reference grid data to network model
- [x] Map onshore H2 nodes to Z2 nodes
- [x] Set economic lifetime to 25y
- [x] Ensure that the radial connections are made directly to the home land node
- [x] Implement TYNDP offshore technology modelling approach
- [x] Set existing capacities
- [x] Address `DEOH002` discrepancy
- [x] Filter data for configured countries
- [x] Fix H2 efficiencies
- [x] Apply PECD capacity factors for offshore wind
- [x] Add a constraint for H2 / DC potentials
- [x] Add a constraint for zone potentials
- [x] Address final `# ToDo` in the code
- [x] Check new files licensing
- [x] Validate myopic approach for existing capacities
- [x] Address warnings
- [x] Add colors for the new carriers
- [x] Remove distribution from Offshore nodes
- [x] Solve infeasibilities
- [x] Create a fig for hubs only
- [x] Add expansion constraint
- [x] Account for time-varying efficiencies across planning horizons (`add_brownfield` and `solve_network`) (see details in https://github.com/open-energy-transition/open-tyndp/pull/68)

## Workflow
1. (new config) `sector:offshore_hubs` defines whether TYNDP offshore hubs must be added to the network.
2. `retrieve_tyndp_bundle` retrieves the raw offshore buses (`tyndp_2024_bundle/Offshore hubs/NODE.xlsx`) and the reference grid data (`tyndp_2024_bundle/Offshore hubs/GRID.xlsx`).
3. (new) `build_tyndp_offshore_hubs` processes the data to create a clean files that define the buses, the grid, the electrolyzers and the generators for the selected scenario and planning horizons.
4. `prepare_sector_network` add the offshore buses and the reference grid to the network.
    - For each node, defines an H2 and AC bus using provided coordinates
    - Add H2 and DC grid connections between hubs, and between hubs and onshore nodes
    - Add power production and H2 production units
    - Add electrolyzers
4. `add_brownfield` adjust offshore expansion by subtracting existing capacity from previous years from current year total capacity and potential
5. `solve_network` introduce two constraints to enforce potentials as defined in input data
    - Add a constraint to limit the expansion of DC and H2 sitting on the same location, as the sum of the two capacities cannot exceed the layer potential
    - Add a constraint to set the maximum potential per zone, taking into account the zone trajectories
6. (new) `plot_base_offshore_network` plot the electricity and hydrogen reference grid for each planning horizon.
7. (new) `plot_offshore_network` plot the electricity and hydrogen expanded grid for each planning horizon.

## Open Issues
- Some nodes have coordinates that are placed on land. Offshore hubs coordinates are only used for plotting purposes, so this is not a major issue.
- Cost assumptions are missing for parts of the reference grid. To address this, any link for which investment assumptions are missing will be considered non-extendable and zero costs.
- No units have been specified for the cost assumptions: for CAPEX, we assume `kEUR/MW`, and for OPEX, assume `kEUR/MW/a`
    - `n.links.capital_cost` default unit is EUR/MW/period. 
        - `capital_cost = (annuity_factor * CAPEX + OPEX) * nyears`
        - We use a 25y economic lifetime (not documented in the methodology but has been confirmed).
- Formatting issues in the tables on p104 of Annex II of the [Methodology report](https://2024.entsos-tyndp-scenarios.eu/wp-content/uploads/2025/01/TYNDP_2024_Scenarios_Methodology_Report_Final_Version_250128.pdf)
- The use and data of `Offshore Hubs/GENERATORS.xlsx` file is not documented. A detailed [notebook](https://github.com/open-energy-transition/open-tyndp/blob/da5dc15b911b594676991c7f9693932b06866e5a/notebooks/test_offshore_potentials.ipynb) has been necessary to validate assumptions.
  - **Existing capacities** will be read from the `LAYER_POTENTIAL` sheet, utilizing technology shares specified in `EXISTING` for hydrogen-generating capacities. A discrepancy of 526 MW for `DEOH002` in 2045 (across all scenarios) is noted when comparing existing capacities with `ZONE_POTENTIAL`. It remains uncertain which of the two values is correct: 5828.55 MW from `LAYER_POTENTIAL` or 6354.55 MW from `ZONE_POTENTIAL`. Currently, the value of 5828.55 MW is used. Additionally, the existing capacity of 3768.25 MW in 2040 exceeds the potential of 3242.25MW shown in the `LAYER_POTENTIAL` sheet.  This value has been adjusted to match the maximum potential value across all scenarios.
  - **Potentials** will be obtained from both the `LAYER_POTENTIAL` and the `ZONE_POTENTIAL` sheets. `LAYER_POTENTIAL` will establish a technology level constraint, while `ZONE_POTENTIAL` will restrict expansion across all technologies at each node. The same 526 MW discrepancy for `DEOH002` in 2045 and 2050 (across all planning horizons and scenarios) has been noted and needs to be addressed to ensure that existing capacities do not exceed their potential. Currently, the value `ZONE_POTENTIAL` value is corrected at 5828.55 MW.
- Why does `GRID.xlsx` contain both capex and opex figures for electricity and hydrogen grid connecting radial nodes to home nodes in both IT and UK? The model currently excludes these candidates since radial assets connect directly to their corresponding home market nodes.
- How are "radial" electrolyzers modeled? Currently, the model assumes these units connect the electricity market node to the corresponding H2 Z2 node.
- The electrical topology obtained from the data differs slightly from the graphical representation in the [Methodology report (section 8.1, p46)](https://2024.entsos-tyndp-scenarios.eu/wp-content/uploads/2025/01/TYNDP_2024_Scenarios_Methodology_Report_Final_Version_250128.pdf). Specifically, the methodology figure omits the link between BEOH001 and DKWOH01, which is present in the dataset. The dataset is assumed to be more accurate where discrepancies exist between data and documentation.

## Notes
The offshore hubs and radial nodes coordinates are presented here below. The radial nodes coordinates are not used as home market coordinates are preferred.
<table>
<tr>
<td width="50%">

**Offshore hubs coordinates**
<img src="https://github.com/user-attachments/assets/a917e18d-4fd6-46b1-859e-c58fc8b13333" width="100%" alt="Offshore hubs coordinates">

</td>
<td width="50%">

**Radial nodes coordinates**
<img src="https://github.com/user-attachments/assets/4d749eeb-33fb-4892-8b46-ae8b62fbfdf4" width="100%" alt="Radial nodes coordinates">

</td>
</tr>
</table>

The 2050 offshore electricity and hydrogen reference grids are presented here below.
<table>
<tr>
<td width="50%">

**Electricity reference grid 2050**
<img src="https://github.com/user-attachments/assets/47fcd1d8-9b78-40b5-ab41-ed19b1b993e1" width="100%" alt="Electricity reference grid 2050">

</td>
<td width="50%">

**Hydrogen reference grid 2050**
<img src="https://github.com/user-attachments/assets/fc5674bb-1659-46fc-ba02-66b3dccf273e" width="100%" alt="Hydrogen reference grid 2050">

</td>
</tr>
</table>

The expanded 2050 offshore electricity and hydrogen grids are presented here below. These graphs are presented to demonstrate the feature, and they do not reflect TYNDP 2024 results.
<table>
<tr>
<td width="50%">

**Expanded 2050 electricity grid**
<img src="https://github.com/user-attachments/assets/a9b7fbe6-c9e5-405d-bed7-2981e14f2a57" width="100%" alt="Expanded 2050 electricity grid">

</td>
<td width="50%">

**Expanded 2050 hydrogen grid**
<img src="https://github.com/user-attachments/assets/756d299d-623d-438e-8e7d-12c5b8b6d2e1" width="100%" alt="Expanded 2050 hydrogen grid">

</td>
</tr>
</table>

The comparison between the offshore network topology and the [Methodology report (section 8.1, p46)](https://2024.entsos-tyndp-scenarios.eu/wp-content/uploads/2025/01/TYNDP_2024_Scenarios_Methodology_Report_Final_Version_250128.pdf) remains relevant. The two topologies are identical, except for the link between BEOH001 and DKWOH01.
<table>
<tr>
<td width="50%">

**TYNDP 2024 topology**
<img src="https://github.com/user-attachments/assets/80473da5-635a-406c-8b30-28245a5bc7c0" width="100%" alt="TYNDP 2024 topology">

</td>
<td width="50%">

**Topology suggested in the PR**
<img src="https://github.com/user-attachments/assets/b828a0c8-193b-49a0-94a3-4058a3ad4e8a" width="100%" alt="Topology suggested in the PR">

</td>
</tr>
</table>



## Checklist

- [x] I tested my contribution locally and it works as intended.
- [x] Code and workflow changes are sufficiently documented.
- [ ] Changed dependencies are added to `envs/environment.yaml`.
- [x] Changes in configuration options are added in `config/config.default.yaml`.
- [x] Changes in configuration options are documented in `doc/configtables/*.csv`.
- [x] Changes in configuration options are added in `config/test/*.yaml`.
- [ ] Sources of newly added data are documented in `doc/data_sources.rst`.
- [x] A release note `doc/release_notes.rst` is added.

daniel-rdt commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,377 @@ def add_gas_and_h2_infrastructure(
+
+    # Load PECD profiles
+    p_max_pu = []
+    for key, fn in profiles.items():
```
Why do we need to iterate over the profiles if we only have `offwind`? Is this to future proof it?

daniel-rdt reviewed (commented):
Thank you @tgilon for all the tedious work on this challenging data integration and the numerous exchanges on the matter in the last days. I like where we landed with this implementation. The current state of the PR is almost good to go in my opinion. 

There are still a few open todos as you also listed in the PR description with the most pressing one the infeasibility from the new constraint. As already discussed I think we should also consider the modelling approach of the hydrogen-generating wind farms to include `p_nom` as `MW_el` and with electrolyser efficiency. @coroa I also tagged you in the corresponding review comment. What do you think?

tgilon commented on config/plotting.default.yaml:
```diff
@@ -275,6 +275,14 @@ plotting:
+    offwind-ac-fl-r: "#a8c5e8"
+    offwind-dc-fb-r: "#b8daf5"
+    offwind-dc-fl-r: "#85b8e6"
+    offwind-dc-fb-oh: "#8fbfec"
+    offwind-dc-fl-oh: "#5a9bd9"
```
I made a new suggestion.

daniel-rdt replied:
looks good

tgilon commented on doc/configtables/sector.csv:
```diff
@@ -229,4 +229,5 @@ imports,,,
--- -- {carrier},currency/MWh,float,Price for importing renewable energy of carrier
\ No newline at end of file
+-- -- {carrier},currency/MWh,float,Price for importing renewable energy of carrier
+offshore_hubs,--,"{true, false}",Add option for TYNDP offshore hubs
```
Done

tgilon commented on scripts/solve_network.py:
```diff
@@ -1112,6 +1112,75 @@ def add_import_limit_constraint(n: pypsa.Network, sns: pd.DatetimeIndex):
+    """
+    ext_i = n.generators.p_nom_extendable
+    gens = n.generators.assign(
+        layer=lambda df: df.index.str.replace(r"-\d{4}$", "-2040", regex=True),
```
Gotcha! That's indeed a bug... Thanks!

tgilon commented on rules/build_sector.smk:
```diff
@@ -1547,6 +1581,26 @@ rule prepare_sector_network:
             if config_provider("sector", "h2_topology_tyndp")(w)
             else []
         ),
+        offshore_buses=lambda w: (
```
Sure, I made the adjustment

tgilon commented on scripts/add_brownfield.py:
```diff
@@ -51,12 +52,14 @@ def add_brownfield(
     # electric transmission grid set optimised capacities of previous as minimum
     n.lines.s_nom_min = n_p.lines.s_nom_opt
-    dc_i = n.links[n.links.carrier == "DC"].index
+    dc_i = n.links[(n.links.carrier == "DC") & (n.links.build_year < year)].index
```
Good catch! This essentially filters out Offshore links, which will be addressed further down in the code. I've fixed the filter as follows:
```python
dc_i = n.links[(n.links.carrier == "DC") & ~(n.links.index.str.contains("Offshore"))].index
```
This should be far more explicit.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,377 @@ def add_gas_and_h2_infrastructure(
+    This function adds offshore generation capacity, various offshore wind
+    turbines (AC and H2), to the offshore hub buses in the network.
+
+    Existing capacities and potentials are adjusted for hydrogen-generating wind farms
```
After researching the feature, it seems that the `efficiency` parameter defined for generators is only used to formulate primary energy related constraints. This wouldn't simplify our formulation.

@coroa Do you agree with keeping it as it is?

tgilon replied:
Ok, perfect. Can you also confirm whether [this](https://github.com/open-energy-transition/open-tyndp/pull/54#issuecomment-3004439593) would make sense?

tgilon replied:
Solved, I've renamed the field to `efficiency_dc_to_b0`.

tgilon commented on scripts/solve_network.py:
```diff
@@ -1112,6 +1112,75 @@ def add_import_limit_constraint(n: pypsa.Network, sns: pd.DatetimeIndex):
+    p_nom = n.model["Generator-p_nom"]
+    lhs = p_nom.loc[dc_gens_i] + p_nom.loc[h2_gens_i] / eff
+
+    existing_l = gens.loc[(h2_i) & ~(ext_i), "p_nom"]
+    grouper_l = gens.loc[h2_i].layer
+    existing_l = existing_l.groupby(grouper_l).sum().reindex(h2_gens_i, fill_value=0)
+    existing_l.index = dc_gens_i
+    rhs = gens.loc[dc_gens_i].p_nom_max - existing_l
```
Good catch! I did it sequentially and did it twice. I'll remove `existing_l` from the constraint formulation.

tgilon commented on scripts/build_tyndp_offshore_hubs.py:
```diff
@@ -0,0 +1,547 @@
+    """
+    Load and process offshore hub coordinates from Excel file.
+
+    Offshore Hubs (OH) nodes are situated offshore, while Offshore Radial (OR) nodes are removed from the data.
```
Correct, this is now done at the end of the script. I'll remove it from this docstring.

tgilon commented on scripts/build_tyndp_offshore_hubs.py:
```diff
@@ -0,0 +1,547 @@
+    This function merges generator data from two sources: existing capacities (EXISTING sheet)
+    and potential capacities (LAYER sheet). It handles reallocation of radial wind farms by
+    correcting inconsistencies between EXISTING and LAYER data, particularly for offshore
+    radial connections that need to be mapped to hub connections.
```
We discussed it in the notebook and also looked at the methodology report  together, see p48.

> To align the existing capacity with the properties specified for offshore wind farms in the PEMMDB, some reallocation of existing capacity between the different configurations was required. Specifically, capacity was shifted to ensure that the share of the existing capacity that is radially connected to the home zone agrees with the data submitted in the PEMMDB. This was also done to align the existing capacity dedicated to hydrogen production.

I cannot say that I fully understood the reallocation mechanism. However, I deduced from the data that existing capacities provided in the `EXISTING` sheet are those collected prior to any reallocations, while the `LAYER_POTENTIAL` sheet is viewed as containing the reallocated existing capacities.

tgilon commented on scripts/_helpers.py:
```diff
@@ -1075,8 +1075,10 @@ def load_cutout(
 
-def make_index(c, carrier):
-    return carrier + " " + c.bus0 + " -> " + c.bus1
+def make_index(c, cname0="bus0", cname1="bus1", prefix="", connector="->", suffix=""):
```
I considered it as well. However, `make_index` only appears twice in the code base, with only one compatible use case. This suggested it provides less general utility than initially expected. But, it's definitely useful for this specific project.

daniel-rdt replied:
Maybe you're right. But it would also only be a minor PR, so probably not a big effort to open it. I could also propose the PR upstream if you don't find the time

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,377 @@ def add_gas_and_h2_infrastructure(
+    mask = offshore_generators["carrier"].str.contains("h2")
+    offshore_generators.loc[mask, "bus"] = offshore_generators.loc[mask, "bus"] + " H2"
+    offshore_generators.index = (
+        offshore_generators.location + " 0 " + offshore_generators.carrier
```
I introduced the bin to minimise changes to `add_brownfield` and ensure compatibility (see [`adjust_renewables_profiles`](https://github.com/PyPSA/pypsa-eur/blob/master/scripts/add_brownfield.py#L225-L242)).

daniel-rdt replied:
I am good with keeping it. It is also not visible in the results but only in the processing code, so won't create any confusion I suppose

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,382 @@ def add_gas_and_h2_infrastructure(
+        bus0=offshore_electrolysers.bus0,
+        bus1=offshore_electrolysers.bus1,
+        p_nom_extendable=True,
+        carrier="H2 Electrolysis",
```
We could use `H2 Offshore Electrolysis`.

tgilon replied:
Yeah, as you can see in the PR, I often need to filter out offshore assets. Currently, I'm using the index for this purpose. I was considering a leaner approach using the carrier. However, as you are mentioned, this would compete with the statistics use case.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,382 @@ def add_gas_and_h2_infrastructure(
+        p_min_pu=offshore_grid_dc.p_min_pu,
+        p_max_pu=offshore_grid_dc.p_max_pu,
+        capital_cost=offshore_grid_dc.capital_cost,
+        carrier="DC",
```
We could use `Offshore DC`.

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,382 @@ def add_gas_and_h2_infrastructure(
+        p_min_pu=offshore_grid_h2.p_min_pu,
+        p_max_pu=offshore_grid_h2.p_max_pu,
+        capital_cost=offshore_grid_h2.capital_cost,
+        carrier="H2 pipeline",
```
We could use `H2 Offshore pipeline`.

tgilon reviewed (commented):
We could make it clearer in the carrier that we are introducing offshore assets.

daniel-rdt commented:
> The current methodology does not account for time-varying efficiencies across different planning horizons. This limitation manifests in two locations: [`add_brownfield`](https://github.com/open-energy-transition/open-tyndp/blob/50b0bf586565d1ebd8d41b5bd5bf98be44753fb5/scripts/add_brownfield.py#L154) and [`solve_network`](https://github.com/open-energy-transition/open-tyndp/blob/50b0bf586565d1ebd8d41b5bd5bf98be44753fb5/scripts/solve_network.py#L1174).
> 
> At both stages, previous electrolysis efficiency values are inaccessible - neither stored in the network nor available as costs. To avoid introducing dependencies on all cost files, the current implementation applies current efficiency. However, this simplification may introduce inconsistencies between planning horizons.
> 
> An alternative approach would store efficiency data during `prepare_sector_network`, when these values are available. `n.generators.efficiency` could therefore be used, but this could interfere with the initial use case of formulating global constraints on primary energy demand emissions.
> 
> What do you think about this, @coroa and @daniel-rdt? Do you have any recommendations?

Could be a viable solution to store the efficiency of the offwind generator-electrolyser combination in the efficiency value of the generator. It could create some confusion as the `p_nom` value is already in MW_H2 and (even though it does not effect it), an efficiency value implies that the generator output would be subject to the efficiency specified.

But if we make sure that the value is not used otherwise and we document thoroughly here it could be a good approach nonetheless. @coroa Do you know how this efficiency value of generators came about?

coroa commented on scripts/solve_network.py:
```diff
@@ -1121,6 +1121,85 @@ def add_import_limit_constraint(n: pypsa.Network, sns: pd.DatetimeIndex):
+    ).rename_axis("Generator-ext")
+
+    # Constraint DC / H2 expansion on the same layer
+    h2_i = gens.carrier.str.contains("h2")
```
can we be more explicit here than anything that contains `h2` ie. a concrete list of the carriers maybe and then `gens.carrier.isin(...)`?

tgilon replied:
Sure, I'll switch it to

```python
    h2_i = gens.carrier.str.startswith("offwind-h2")
```

tgilon replied:
I made it more explicit. See https://github.com/open-energy-transition/open-tyndp/pull/54/commits/9926a23984b659748bc27ce55dcf974f4bfe58d9

coroa commented on scripts/solve_network.py:
```diff
@@ -1121,6 +1121,85 @@ def add_import_limit_constraint(n: pypsa.Network, sns: pd.DatetimeIndex):
+        .p_nom_max
+    )
+
+    off_i = gens.index.str.contains("offwind")
```
should this be on the carrier and maybe also isin based?

tgilon replied:
I made it more explicit. See https://github.com/open-energy-transition/open-tyndp/pull/54/commits/9926a23984b659748bc27ce55dcf974f4bfe58d9

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,392 @@ def add_gas_and_h2_infrastructure(
+        Technology costs assumptions.
+    spatial : object, optional
+        Object containing spatial information about nodes and their locations.
+    logger : logging.Logger, optional
+        Logger for output messages. If None, no logging is performed.
```
See above

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,392 @@ def add_gas_and_h2_infrastructure(
+
+    # Adjust capacities and costs to account for efficiency
+    h2_idx = offshore_generators.filter(like="h2", axis=0).index
+    offshore_generators.loc[h2_idx, ["p_nom_min", "p_nom_max"]] = (
+        offshore_generators.loc[h2_idx, ["p_nom_min", "p_nom_max"]].mul(
+            costs.at["electrolysis", "efficiency"]
+        )
+    )
```
```suggestion
    offshore_generators.loc[h2_idx, ["p_nom_min", "p_nom_max"]] *=  costs.at["electrolysis", "efficiency"]
```

Probably works too

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,392 @@ def add_gas_and_h2_infrastructure(
+            costs.at["electrolysis", "efficiency"]
+        )
+    )
+    offshore_generators.loc[h2_idx, ["capex", "opex"]] = offshore_generators.loc[
+        h2_idx, ["capex", "opex"]
+    ].div(costs.at["electrolysis", "efficiency"])
```
`/=` ?

coroa reviewed (commented):
Did only have very limited time to look at this and the comments are more a draft state. I'll read it for real on monday or tuesday. :)

tgilon commented on scripts/prepare_sector_network.py:
```diff
@@ -3010,6 +3031,392 @@ def add_gas_and_h2_infrastructure(
+        e.g. {'offwind-dc': 'path/to/profile.nc'}
+    costs : pd.DataFrame
+        Technology costs assumptions.
+    logger : logging.Logger
+        Logger for output messages. If None, no logging is performed.
```
There's no good reason for it, so I'll remove it.

tgilon commented on config/config.default.yaml:
```diff
@@ -810,6 +810,11 @@ sector:
       oil: 125
+  offshore_hubs_tyndp:
+    enable: false
+    max_capacity:
+      DC: 10
+      H2: 30
```
Done

tgilon commented on rules/postprocess.smk:
```diff
@@ -83,7 +110,7 @@ if config["foresight"] != "perfect":
         output:
             map=RESULTS
-            + "maps/base_s_{clusters}_{opts}_{sector_opts}-h2_network_{planning_horizons}.pdf",
+            + "maps/base_s_{clusters}_{opts}_{sector_opts}_{planning_horizons}-h2_network.pdf",
```
Indeed, this makes the set of outputs more coherent

```
base_s_all___2030-balance_map_AC.pdf
base_s_all___2030-balance_map_co2 stored.pdf
base_s_all___2030-balance_map_gas.pdf
base_s_all___2030-balance_map_H2.pdf
base_s_all___2030-balance_map_methanol.pdf
base_s_all___2030-balance_map_oil.pdf
base_s_all___2030-h2_network.pdf
base_s_all___2030-offshore_network_DC.pdf
base_s_all___2030-offshore_network_H2.pdf
```

coroa commented on rules/build_sector.smk:
```diff
@@ -1291,6 +1291,18 @@ def input_heat_source_power(w):
+    ]
+    if config_provider("sector", "offshore_hubs_tyndp", "enable")(w):
+        return {f: resources(f"{f}.csv") for f in offshore_files}
+    return {f: [] for f in offshore_files}
```
i'd expect that it is fine to just return an empty dictionary

tgilon replied:
Any other takeaways here before merging?

tgilon replied:
Done

coroa commented on rules/build_sector.smk:
```diff
@@ -1354,6 +1366,40 @@ if config["sector"]["h2_topology_tyndp"]:
+            countries=config_provider("countries"),
+            offshore_hubs_tyndp=config_provider("sector", "offshore_hubs_tyndp"),
+        input:
+            nodes=directory("data/tyndp_2024_bundle/Offshore hubs/NODE.xlsx"),
+            grid=directory("data/tyndp_2024_bundle/Offshore hubs/GRID.xlsx"),
+            electrolysers=directory(
+                "data/tyndp_2024_bundle/Offshore hubs/ELECTROLYSER.xlsx"
+            ),
+            generators=directory("data/tyndp_2024_bundle/Offshore hubs/GENERATOR.xlsx"),
```
What are the `directory` flags for? Aren't those output only ones?

coroa commented on rules/solve_electricity.smk:
```diff
@@ -11,8 +11,14 @@ rule solve_network:
+        carriers_tyndp=config_provider("electricity", "tyndp_renewable_carriers"),
     input:
         network=resources("networks/base_s_{clusters}_elec_{opts}.nc"),
+        offshore_zone_trajectories=lambda w: (
+            resources("offshore_zone_trajectories.csv")
+            if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+            else []
+        ),
```
branch

coroa commented on rules/solve_overnight.smk:
```diff
@@ -11,10 +11,16 @@ rule solve_sector_network:
         network=resources(
             "networks/base_s_{clusters}_{opts}_{sector_opts}_{planning_horizons}.nc"
         ),
+        offshore_zone_trajectories=lambda w: (
+            resources("offshore_zone_trajectories.csv")
+            if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+            else []
+        ),
```
branch

coroa commented on rules/solve_perfect.smk:
```diff
@@ -107,11 +101,17 @@ rule solve_sector_network_perfect:
             "networks/base_s_{clusters}_{opts}_{sector_opts}_brownfield_all_years.nc"
         ),
         costs=resources("costs_2030.csv"),
+        offshore_zone_trajectories=lambda w: (
+            resources("offshore_zone_trajectories.csv")
+            if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+            else []
+        ),
```
branch

coroa commented on scripts/prepare_sector_network.py:
```diff
@@ -1630,7 +1649,9 @@ def insert_electricity_distribution_grid(
     """
 
-    nodes = n.buses.query("carrier == 'AC' and not index.str.contains('DRES')").index
+    nodes = n.buses.query(
+        "carrier == 'AC' and not index.str.contains('DRES') and not index.str.contains('OH')"
+    ).index
```
That is a very nitty gritty intervention, any idea how we can make the filter here more robust which works for upstream in general?

daniel-rdt replied:
For upstream I introduced this PR [here](https://github.com/PyPSA/pypsa-eur/pull/1672/files) which basically takes into account all the AC buses instead of the pop_layout.index for creating the lv buses

tgilon replied:
What about using the `type` field? This field is already populated for Offshore Hubs and could easily be populated for `DRES` as well. This would simplify the expression to:

```suggestion
    nodes = n.buses.query("carrier == 'AC' and type==''").index
```

coroa reviewed (commented):
Soo, i finally made it through. Let's schedule a meeting (tomorrow?) to discuss the remaining todos and what to do about those efficiencies?

tgilon commented on rules/solve_myopic.smk:
```diff
@@ -142,11 +131,17 @@ rule solve_sector_network_myopic:
             "networks/base_s_{clusters}_{opts}_{sector_opts}_{planning_horizons}_brownfield.nc"
         ),
         costs=resources("costs_{planning_horizons}.csv"),
+        offshore_zone_trajectories=lambda w: (
+            resources("offshore_zone_trajectories.csv")
+            if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+            else []
+        ),
```
We address them all

tgilon commented on scripts/build_tyndp_offshore_hubs.py:
```diff
@@ -0,0 +1,561 @@
+        "National Trends": "NT",
+    }
+
+    # Load reference grid
+    grid = pd.read_excel(
+        fn,
+        sheet_name="Reference grid",
+    ).rename(columns=column_dict)
+    grid["carrier"] = grid["carrier"].replace("E", "DC")
+    grid = expand_all_scenario(grid, scenario_dict.values()).query(
+        "scenario == @scenario"
+    )
```
Correct, but that is indeed the case. The data only contains "all". Let's generalise anyway.

tgilon commented:
@coroa @daniel-rdt I've addressed the todos. Below is a summary of the changes:
- Updated carrier naming to distinguish offshore hubs from other components
- Fixed `max_capacity` condition to apply only to inter-hubs connections
- Added plotting threshold parameter to `plot_offshore_network`
- Fixed other minors issues (see commits)

coroa commented:
Feel free to merge once they are addressed

tgilon commented on Snakefile:
```diff
@@ -78,9 +77,56 @@ if config["foresight"] == "perfect":
+def input_all_tyndp(w):
+    files = []
+    files.extend(
+        expand(
+            (
+                resources(
+                    "maps/base_h2_network_{clusters}_{opts}_{sector_opts}_{planning_horizons}.pdf"
+                )
+                if config_provider("sector", "H2_network")(w)
+                else []
+            ),
+            run=config["run"]["name"],
+            **config["scenario"],
+        )
+    )
+    files.extend(
+        expand(
+            (
+                resources(
+                    "maps/base_offshore_network_{clusters}_{opts}_{sector_opts}_{planning_horizons}_{carrier}.pdf"
+                )
+                if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+                else []
+            ),
+            run=config["run"]["name"],
+            **config["scenario"],
+            carrier=config_provider("plotting", "offshore_maps", "bus_carriers")(w),
+        )
+    )
+    files.extend(
+        expand(
+            (
+                RESULTS
+                + "maps/base_s_{clusters}_{opts}_{sector_opts}_{planning_horizons}-offshore_network_{carrier}.pdf"
+                if config_provider("sector", "offshore_hubs_tyndp", "enable")(w)
+                else []
+            ),
+            run=config["run"]["name"],
+            **config["scenario"],
+            carrier=config_provider("plotting", "offshore_maps", "bus_carriers")(w),
```
Done

daniel-rdt commented on scripts/add_brownfield.py:
```diff
@@ -128,15 +129,30 @@ def add_brownfield(
             )
 
             # account for the shared potential of hydrogen- and electricity-generating wind farms
-            already_existing_l = already_existing.copy()
```
You removed the redundancy. I do prefer this however, over the `else` clause. If I am not missing anything, you could still assign this here, couldn't you?
```
already_existing_l = already_existing.p_nom_opt
```
I am also fine if you decide to keep it as is, since it is minor

tgilon replied:
I believe it's clearer since `already_existing_l` is overwritten rather than modified within the `if` clause
